### DISCUSSION TITLE
lsp-zero with Lazy package manager issues

### DISCUSSION THREAD

**Discussion by aryklein** [VonHeikemen/lsp-zero.nvim#147] [Q&A] [answered]
I switched from Packer to Lazy and I have issues. First thing I did is to define the plugin in Lazy by:

```lua
{
  'VonHeikemen/lsp-zero.nvim',
  branch = 'v1.x',
  dependencies = {
    -- LSP Support
    {'neovim/nvim-lspconfig'},             -- Required
    {'williamboman/mason.nvim'},           -- Optional
    {'williamboman/mason-lspconfig.nvim'}, -- Optional

    -- Autocompletion
    {'hrsh7th/nvim-cmp'},         -- Required
    {'hrsh7th/cmp-nvim-lsp'},     -- Required
    {'hrsh7th/cmp-buffer'},       -- Optional
    {'hrsh7th/cmp-path'},         -- Optional
    {'saadparwaiz1/cmp_luasnip'}, -- Optional
    {'hrsh7th/cmp-nvim-lua'},     -- Optional

    -- Snippets
    {'L3MON4D3/LuaSnip'},             -- Required
    {'rafamadriz/friendly-snippets'}, -- Optional
  }
}
```

and then I required the module in the file `.config/nvim/lua/ary/lsp-zero.lua`:

```lua
-- Learn the keybindings, see :help lsp-zero-keybindings
-- Learn to configure LSP servers, see :help lsp-zero-api-showcase
local lsp = require('lsp-zero')
lsp.preset('recommended')

-- only for lskind-nvim
-- See https://github.com/VonHeikemen/lsp-zero.nvim/discussions/143#discussioncomment-4845982
-- block taken from here: https://github.com/onsails/lspkind.nvim#option-2-nvim-cmp
lsp.setup_nvim_cmp({
    formatting = {
        fields = {'abbr', 'kind', 'menu'},
        format = require('lspkind').cmp_format({
            mode = 'symbol_text', -- show only symbol annotations
            maxwidth = 50, -- prevent the popup from showing more than provided characters (e.g 50 will not show more than 50 characters)
            ellipsis_char = '...', -- when popup menu exceed maxwidth, the truncated part would show ellipsis_char instead (must define maxwidth first)
        })
    }
})

lsp.nvim_workspace()
lsp.setup()

-- diagnostic config
vim.diagnostic.config({
    virtual_text = true,
    signs = true,
    update_in_insert = false,
    underline = true,
    severity_sort = false,
    float = {
        focusable = false,
        style = 'minimal',
        border = 'rounded',
        source = 'always',
        header = '',
        prefix = '',
    },
})
```

But LSP and auto completion won't work. 

I guess it's something related to async code loaded, so I put the code here `.config/nvim/after/plugin/lsp-zero.lua` and worked.

My next try (that worked) was to define everything like this:

```lua
return {
  'VonHeikemen/lsp-zero.nvim',
  branch = 'v1.x',
  dependencies = {
    -- LSP Support
    {'neovim/nvim-lspconfig'},             -- Required
    {'williamboman/mason.nvim'},           -- Optional
    {'williamboman/mason-lspconfig.nvim'}, -- Optional

    -- Autocompletion
    {'hrsh7th/nvim-cmp'},         -- Required
    {'hrsh7th/cmp-nvim-lsp'},     -- Required
    {'hrsh7th/cmp-buffer'},       -- Optional
    {'hrsh7th/cmp-path'},         -- Optional
    {'saadparwaiz1/cmp_luasnip'}, -- Optional
    {'hrsh7th/cmp-nvim-lua'},     -- Optional

    -- Snippets
    {'L3MON4D3/LuaSnip'},             -- Required
    {'rafamadriz/friendly-snippets'}, -- Optional

    -- lspkind
    {'onsails/lspkind-nvim'},
  },
  config = function()
      local lsp = require('lsp-zero')
      lsp.preset('recommended')

      -- only for lskind-nvim
      -- See https://github.com/VonHeikemen/lsp-zero.nvim/discussions/143#discussioncomment-4845982
      -- block taken from here: https://github.com/onsails/lspkind.nvim#option-2-nvim-cmp
      lsp.setup_nvim_cmp({
          formatting = {
              fields = {'abbr', 'kind', 'menu'},
              format = require('lspkind').cmp_format({
                  mode = 'symbol_text', -- show only symbol annotations
                  maxwidth = 50, -- prevent the popup from showing more than provided characters (e.g 50 will not show more than 50 characters)
                  ellipsis_char = '...', -- when popup menu exceed maxwidth, the truncated part would show ellipsis_char instead (must define maxwidth first)
              })
          }
      })

      lsp.nvim_workspace()
      lsp.setup()

      -- diagnostic config
      vim.diagnostic.config({
          virtual_text = true,
          signs = true,
          update_in_insert = false,
          underline = true,
          severity_sort = false,
          float = {
              focusable = false,
              style = 'minimal',
              border = 'rounded',
              source = 'always',
              header = '',
              prefix = '',
          },
      })
 end
}
```

Do you know what could be the issue with the approach of defining the plugin on Lazy and then requiring the config in another file?

**Comment by VonHeikemen** [comment]
Try adding `lazy = false` as a property.

**Reply by aryklein** [reply]
I tried this option before and changing priorities but didn't help. When executing `LsPInfo` I see issues with `lspconfig-root-detection` :

![2023-02-07_17-06-06](https://user-images.githubusercontent.com/8699850/217353866-53a5934b-a035-4c60-9c79-e872a5afbcb1.png)

**Reply by VonHeikemen** [reply]
You should still disable anything related to lazy loading. The plugin `lspconfig` does weird things when is not loaded at startup.

**Comment by aryklein** [comment]
To add more context, sometimes it works without changing the config. It seems to be related to load timings. When it works I get:
```
root directory:  Running in single file mode.
```

**Reply by VonHeikemen** [reply]
I think the trick is to load it before Neovim sets the filetype for the current file. But I recommend setting lazy.nvim to run lsp-zero at startup.

**Reply by aryklein** [reply]
I did that but not luck! I will keep you posted if I find the issue.
Thanks anyway.

**Reply by aryklein** [reply]
https://github.com/VonHeikemen/lsp-zero.nvim/pull/148
That worked for me. I hope it helps

**Reply by aryklein** [reply]
This issue seems to be affecting other plugins: https://github.com/folke/lazy.nvim/discussions/369

**Reply by folke** [reply]
@aryklein that has nothing to do with your issue here. What you linked to, is chemoiz that overrides the default Neovim ftplugin detection code. That's just bad practice and I'm not willing to support those types of use-cases. There are far better ways to achieve what that plugin does.

Your second approach (doing lspzero setup in the config function) is how you should do it and it seems it does work like that from reading your comment?

I don't use lsp-zero, but I do use lspconfig, mason etc... And I can easily lazy-load those plugins.

**Reply by aryklein** [reply]
Oh, thanks to clarify my mistake in the comment here. Thanks @folke!
I solved it setting the highest priority.

**Comment by aryklein** [comment, marked as answer]
@VonHeikemen I found the way to reproduce the issue. It only happens when nvim-treesitter is loaded before lspconfig plugin. I have no idea why. You only need to make sure lsp-zero is loaded before nvim-treesitter

**Reply by kskarthik** [reply]
i also have the same issue, how did you fix (code snippet please) ?

**Reply by VonHeikemen** [reply]
@kskarthik did you tried using the snippet in the section [Lazy loading with lazy.nvim](https://github.com/VonHeikemen/lsp-zero.nvim/blob/v2.x/doc/md/guides/lazy-loading-with-lazy-nvim.md)?

**Reply by aryklein** [reply]
@kskarthik  I managed to solve it by using `priority` and forcing it to load right after the color scheme:

```lua
return {
  'VonHeikemen/lsp-zero.nvim',
  branch = 'v1.x',
  lazy = false,
  priority = 51,
  dependencies = {
    -- LSP Support
    {'williamboman/mason.nvim'},           -- Optional
    {'williamboman/mason-lspconfig.nvim'}, -- Optional
    {'neovim/nvim-lspconfig'},             -- Required

    -- Autocompletion
    {'hrsh7th/nvim-cmp'},         -- Required
    {'hrsh7th/cmp-nvim-lsp'},     -- Required
    {'hrsh7th/cmp-buffer'},       -- Optional
    {'hrsh7th/cmp-path'},         -- Optional
    {'saadparwaiz1/cmp_luasnip'}, -- Optional
    {'hrsh7th/cmp-nvim-lua'},     -- Optional

    -- Snippets
    {'L3MON4D3/LuaSnip'},             -- Required
    {'rafamadriz/friendly-snippets'}, -- Optional

    -- lspkind
    {'onsails/lspkind.nvim'}, -- Optional'},
  },
  config = function()
      require('ary.lsp-zero')
  end
}
```

Anyway I will try the snippet by @VonHeikemen. I just saw it :)

**Reply by kskarthik** [reply]
@aryklein thank you for the snippet! I will try it. BTW what does 51 mean in priority key

**Reply by VonHeikemen** [reply]
Keep in mind the documentation assumes you are using version 2 of lsp-zero. This question was for version 1.

**Reply by kskarthik** [reply]
@VonHeikemen yeah, i got it. Just found about priority, 50 is the default

![image](https://user-images.githubusercontent.com/11899221/232551283-7b20c5e4-1053-432a-b44d-14937f46e1fb.png)
